src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs

```diff
@@ -282,6 +196,140 @@ await using (sslDuplexPipe)
+
+        internal static bool RemoteCertificateValidationCallback(
+            ClientCertificateMode clientCertificateMode,
+            Func<X509Certificate2, X509Chain, SslPolicyErrors, bool> clientCertificateValidation,
+            X509Certificate certificate,
+            X509Chain chain,
+            SslPolicyErrors sslPolicyErrors)
+        {
+            if (certificate == null)
+            {
+                return clientCertificateMode != ClientCertificateMode.RequireCertificate;
+            }
+
+            if (clientCertificateValidation == null)
+            {
+                if (sslPolicyErrors != SslPolicyErrors.None)
+                {
+                    return false;
+                }
+            }
+
+            var certificate2 = ConvertToX509Certificate2(certificate);
+            if (certificate2 == null)
+            {
+                return false;
+            }
+
+            if (clientCertificateValidation != null)
+            {
+                if (!clientCertificateValidation(certificate2, chain, sslPolicyErrors))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private SslDuplexPipe CreateSslDuplexPipe(IDuplexPipe transport, MemoryPool<byte> memoryPool)
+        {
+            var inputPipeOptions = new StreamPipeReaderOptions
+            (
+                pool: memoryPool,
+                bufferSize: memoryPool.GetMinimumSegmentSize(),
+                minimumReadSize: memoryPool.GetMinimumAllocSize(),
+                leaveOpen: true
+            );
+
+            var outputPipeOptions = new StreamPipeWriterOptions
+            (
+                pool: memoryPool,
+                leaveOpen: true
+            );
+
+            return new SslDuplexPipe(transport, inputPipeOptions, outputPipeOptions, _sslStreamFactory);
+        }
+
+        private static async ValueTask<SslServerAuthenticationOptions> ServerOptionsCallback(SslStream stream, SslClientHelloInfo clientHelloInfo, object state, CancellationToken cancellationToken)
+        {
+            var (middleware, context, feature) = (ValueTuple<HttpsConnectionMiddleware, ConnectionContext, Core.Internal.TlsConnectionFeature>)state;
+
+            feature.HostName = clientHelloInfo.ServerName;
+
+            var sslOptions = await middleware._serverOptionsSelectionCallback(stream, clientHelloInfo, middleware._serverOptionsSelectionCallbackState, cancellationToken);
+
+            // REVIEW: Cache results? We don't do this for ServerCertificateSelectionCallback.
+            if (sslOptions.ServerCertificate is X509Certificate2 cert)
+            {
+                EnsureCertificateIsAllowedForServerAuth(cert);
+            }
+
+            // REVIEW: Should we write any event before this?
+            KestrelEventSource.Log.TlsHandshakeStart(context, sslOptions);
```

Comment:
There are two branches where we start the handshake. Since we don't have the `sslOptions` when we start the callback-based version of AuthenticateAsServerAsync, we log it in the callback itself. I now have a comment in the old HttpsConnectionMiddlware about how to completely consolidate these two branches (we already share methods where possible). I don't want to do this until https://github.com/dotnet/runtime/issues/40402 is resolved though, since it would break a lot of our client cert logic.